Rewrite raspberrypi-builder.sh based on extlinux

jbaum98 opened the pull request:
This adds the ability to set a configuration limit for the raspberry pi
firmware as well, so we can change the
boot.loader.raspberryPi.uboot.configurationLimit option to
boot.loader.raspberryPi.configurationLimit and use it for both uboot and
the raspberry pi firmware.

It also fixes a bug where not everything in /boot/old would be cleaned
up.

In addition, this passes shellcheck with no errors, and runs shellcheck
as a part of the build.

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!
List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://hydra.nixos.org/job/nixpkgs/trunk/manual/latest/download/1/nixpkgs/manual.html#chap-reviewing-contributions
-->

###### Motivation for this change


###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

jbaum98 committed:
Rewrite raspberrypi-builder.sh based on extlinux

This adds the ability to set a configuration limit for the raspberry pi
firmware as well, so we can change the
boot.loader.raspberryPi.uboot.configurationLimit option to
boot.loader.raspberryPi.configurationLimit and use it for both uboot and
the raspberry pi firmware.

It also fixes a bug where not everything in /boot/old would be cleaned
up.

In addition, this passes shellcheck with no errors, and runs shellcheck
as a part of the build.

Co-authored-by: Samuel Dionne-Riel <samuel@dionne-riel.com>

samueldr commented on nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.sh:
```diff
@@ -1,4 +1,4 @@
-#! @bash@/bin/sh
+#! @bash@/bin/bash -e
```
In #90119 it was decided to remove `-e` which is redundant because the `set -e` coming right after. (See the resolved discussion I can't link to because it's resolved in that PR.)

In some situations the script ends up being called in a way where the shebang is disregarded.

jbaum98 replied:
Ah I saw this, but it must have gotten clobbered when I copied over the extlinux script. Should I go in and fix it in the extlinux script too while I'm at it?

samueldr replied:
No need to, it have to be checked whether it's running with or without `-e` enabled properly beforehand, and it expands the scope of this PR needlessly.

samueldr commented on nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.nix:
```diff
@@ -3,6 +3,9 @@
   src = ./raspberrypi-builder.sh;
   isExecutable = true;
+  postInstall = "shellcheck $out";
+  nativeBuildInputs = [pkgs.shellcheck];
```
I need to verify, but I recall that `nativeBuildInputs` and splicing won't work during cross-compilation when done like here into a NixOS module's expression.

The solution is to use `pkgs.buildPackages` rather than `pkgs`.

```suggestion
  nativeBuildInputs = [pkgs.buildPackages.shellcheck];
```

samueldr commented:
I would still need to read the full script, and test it, not sure when I personally can. If another contributor/committer does, it's fine too.

I guess one last thing to do, once everything is done and tested, would be to squash the commits.

Thanks for the fixes btw!

jbaum98 commented:
Great! Thanks for your review!

Yes, it's working for me on my pi but it'd be great if someone else could test it.

Luflosi commented:
What do I need to do to test this on my Pi 4?

jbaum98 commented:
You basically need to use this pull request for your `system/boot/loader/raspberrypi/raspberrypi.nix`. So maybe add a module like this to your NixOS configuration.
```nix
let nixpkgs_rpi_loader = builtins.fetchTarball {
  url = "https://github.com/jbaum98/nixpkgs/archive/68d9a30b4ddb52cbcb98eae11a14fd7d0dd945f2.tar.gz";
  sha256 = "1vglz99gaffsprzkpnhpafxhd5ckwxhn3sfz87lqp3h6q0h6xblw";
};
in {
  disabledModules = [ "system/boot/loader/raspberrypi/raspberrypi.nix" ];
  imports = [ "${nixpkgs_rpi_loader}/nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix" ];
}
```
That's not exactly [what I have](https://git.sr.ht/~jakewaksbaum/pi/tree/275bd4414fc6f2c579df6ab3b3a7887ce75f70a1/flake.nix#L20-23) because I'm using flakes, but I think it should work.

Then, if you want, you can set `boot.loader.raspberryPi.configurationLimit` to something lower than it's default value, 20. It should then, when you run `nixos-rebuild switch` or `/run/current-system/bin/switch-to-configuration activate`, delete the oldest configurations from the `boot` partition until you have that number.

golddranks commented:
Thanks for the instructions. I tried this on my Pi 4. Seems to work! I was having troubles with /boot being full all the time, and this also seems to fix that.

Luflosi commented:
I'd like to thank you for the instructions as well. Unfortunately I didn't get a chance to test this yet as my SD card broke but I should be getting a new one in a couple days, then I can finally test this.

Luflosi commented:
After I got a new SD card, I was now able to test this and it seems to work just fine (as expected). Thank you very much for this PR.

jbaum98 commented:
@samueldr I squashed the commits (but I made a new branch pointing to the old commits to prevent GC, so hopefully the snippet above will still work). Let me know if there's anything else I can do to help get this merged.

golddranks commented:
Any movement around getting this merged?

golddranks commented:
I'm trying to solve the shellcheck/pandoc problem here: https://github.com/NixOS/nixpkgs/issues/96154

golddranks commented:
Once https://github.com/NixOS/nixpkgs/pull/96272 gets fixed, the problem I was encountering with this PR should be fixed.

golddranks commented:
@samueldr @tkerber If this isn't merged soon, it will miss the 20.09 release, which would be a shame – Raspberry Pi 4 is an inch away from working flawlessly out of the box. The blocker I mentioned earlier is merged, so the build should work now. Unfortunately, I haven't been able to re-test it at the moment, because the unstable channel seems to have been blocked for days, and the unstable-small channel builds a lot of stuff from source which takes absolutely forever to on my Raspberry Pi 4... I wonder if there's any chance to get this to the release?

numinit commented:
I can also confirm that the shellcheck step seems like it breaks this PR on aarch64.

numinit commented:
Here's the error... look familiar to anyone?

```
Linking Setup ...
/nix/store/2ns4qjpj9hp7y18wcag909vm81y6wyan-binutils-2.31.1/bin/ld.gold: out of memory
collect2: error: ld returned 1 exit status

out of memory allocating 52864 bytes after a total of 18446743524305431976 bytes
collect2: error: ld returned 1 exit status
`cc' failed in phase `Linker'. (Exit code: 1)
`cc' failed in phase `Linker'. (Exit code: 1)
builder for '/nix/store/fq6kbw9a37syv0bcsc4cd1gpi24z71bg-trifecta-2.1.drv' failed with exit code 1
```

numinit commented:
Maybe in the interest of hitting the merge window, we can temporarily get rid of the shellcheck step?

jbaum98 commented:
Yeah, to be honest now I'm kind of regretting that srep. I was very worried about getting this stuff right, so I appreciated the help and figured it was good to encode the fact that it passed shellcheck in the code, and to make sure noone caused a regression there, but now it kind of seems like more trouble than it's worth?

numinit commented:
Yeah, definitely appreciate the intention. (Un)fortunately Shellcheck is a bunch of Haskell, which leads to dependencies of dependencies that increases the build complexity on aarch64...

BTW, can you update the checkboxes in the PR description? Might have a better chance of merge with that.

golddranks commented:
I agree that Shellcheck is a nice to have as a part of the CI process, but the amount and weight of the additional dependencies is concerning. My Rpi4 is still compiling pandoc after a day, and I'm not sure if it's stuck or not. If there would be a guaranteed binary distribution, this wouldn't be necessarily as concerning, but I'm not familiar about what level of binary distribution is supported, especially for AArch64. Also, I don't see Shellcheck as something that would need to run as a part of the build on every user's machine, but more as a CI tool that prevents oopsies. Is there any distinction with that in NixOS?

golddranks commented:
I tried to re-test this patch, this time changing strategy: I'm now cross-compiling the needed dependencies and then copy them to the Raspberry Pi 4. However, I'm encountering some problems I elaborate in the main Raspberry Pi 4 support thread: https://github.com/NixOS/nixpkgs/issues/63720#issuecomment-685695090

@jbaum98  It might be easier if you'd just disable shellcheck for now.

jbaum98 commented:
Let me see if I can somehow move it to checkPhase so that it can maybe more easily be disabled without removing it entirely.

golddranks commented:
I finally managed to build a working shellcheck for my system, and confirmed this PR to work again.

Here's some tips for other testers:
- It seems that the binaries are now cached, but the unstable channel is still stuck. However, nixpkgs-unstable is not. If you do `nix build -f channel:nixpkgs-unstable pkgs.shellcheck`, it will be able to realise shellcheck in your store, which should allows you test this!
- In case there are people struggling with long build times or too little of memory on Raspberry Pi 4, I found it quite easy to build on an AWS Graviton instance and `nix-copy-closure` the build results to RPi4. Tried with instance type `a1.xlarge` and image `NixOS-20.03.2351.f8248ab6d9e-aarch64-linux`.

Jaculabilis commented:
I was able to get this to work on my rpi3 and my rpi4, but only by copying the changed files locally and then commenting out the two lines about shellcheck in `raspberrypi-builder.nix`. If it's not absolutely necessary to make the other changes function, can we remove it and try it again in a later PR? I'd love to see the other changes here merged, and I'd hate for it to be blocked over something that can be done later.

jbaum98 committed:
Move shellcheck to checkPhase and disable check

There seem to be issues with ghc on aarch64, so this makes it easy to
enable/disable the use of shellcheck, which is written in Haskell.

Jaculabilis commented:
Tested via executing:
```
mkdir -p /etc/nixos/modules/system/boot/loader/raspberrypi/
wget 'https://raw.githubusercontent.com/NixOS/nixpkgs/0ed97a81168b3c725f8a6555d2f005bd62709ca2/nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.nix' -O /etc/nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.nix
wget 'https://raw.githubusercontent.com/NixOS/nixpkgs/0ed97a81168b3c725f8a6555d2f005bd62709ca2/nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.sh' -O /etc/nixos/modules/system/boot/loader/raspberrypi/raspberrypi-builder.sh
wget 'https://raw.githubusercontent.com/NixOS/nixpkgs/0ed97a81168b3c725f8a6555d2f005bd62709ca2/nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix' -O /etc/nixos/modules/system/boot/loader/raspberrypi/raspberrypi.nix
```
then adding to `/etc/nixos/configuration.nix`:
```
  disabledModules = [ "system/boot/loader/raspberrypi/raspberrypi.nix" ];
  imports = [ ./modules/system/boot/loader/raspberrypi/raspberrypi.nix ];
```
then `nixos-rebuild build`. The build works fine and I didn't see it fetch a ton of Haskell, so I think the change worked.

jbaum98 commented:
I also tested by overriding an input in my flake and it too didn't pull in haskell.

numinit commented:
FYI: you might have to rebase against #97462 when it's merged.

jbaum98 commented:
Just to make sure I get it: we still want shellcheck to be from `buildPackages` because checking happens on the builder not the target?

numinit commented:
Yes, that makes sense to me.

numinit commented:
Another PR that will be of interest: #97883

Looks like the rpi4 is getting proper u-boot support so the images can be closer to the standard aarch64 images. I still think the option rename might be a good idea though.

jbaum98 commented:
Closing now that we have uboot support for the rpi4.
